src/resolver/amazon/amazon-resolver.service.ts

```diff
@@ -77,10 +77,14 @@ export class AmazonResolverService extends AbstractResolver {
       // main info
       message.setTitle(HtmlUtil.getTextContent(title as HTMLElement));
       message.setAuthor(HtmlUtil.getTextContent(author as HTMLElement));
-      this.setDetails(message, siteLanguage, details as HTMLElement);
+      const asinFromDetails: string | null = this.setDetails(
```

Comment:
Change this method to `getDetails(siteLanguage: string, details: HTMLElement): AmazonDetails`
AmazonDetails should have `publisher: string | null, language: string | null, asin: string | null`
Check that publisher and language have values and set them to the message object.
Use the asin in `getAsin`, but change it to get the asin from URL only if asin from detail is null.